src/input/mod.rs

```diff
@@ -1139,7 +1140,93 @@ impl State {
                     );
                 }
             }
-            _ => { /* TODO e.g. tablet or touch events */ }
+            InputEvent::TouchDown { event, .. } => {
+                if let Some(seat) = self.common.seat_with_device(&event.device()).cloned() {
+                    let current_output = seat.active_output();
+                    let geometry = current_output.geometry();
+
+                    // XXX
+                    let position = geometry.loc.to_f64()
+                        + event
+                            .position_transformed(geometry.size.as_logical())
+                            .as_global();
+
+                    let overview = self.common.shell.overview_mode();
+                    let workspace = self.common.shell.workspaces.active_mut(&current_output);
+                    let under = State::surface_under(
+                        position,
+                        &current_output,
+                        &self.common.shell.override_redirect_windows,
+                        overview.0.clone(),
+                        workspace,
+                        self.common.session_lock.as_ref(),
+                    )
+                    .map(|(target, pos)| (target, pos.as_logical()));
+
+                    if let Some((target, pos)) = under {
+                        if let Some(wl_surface) = target.wl_surface() {
+                            let serial = SERIAL_COUNTER.next_serial();
+                            let mut touch = seat.get_touch().unwrap();
```

Comment:
We should check here, if the client has actually bound wl_touch and do pointer emulation otherwise.

Reply:
Do any other compositors emulate `wl_pointer` for touch events like that? Or is it generally assumed clients are responsible for that? That should be safe to do though, for any clients that don't.

Reply:
Honestly I am not sure. I never looked into this a lot, but at least Xwayland will never bind wl_touch, so we need to do emulation for some clients anyway or there is no way to interact with them.

Reply:
I haven't tested it, but https://gitlab.freedesktop.org/xorg/xserver/-/blob/master/hw/xwayland/xwayland-input.c does call `wl_seat_get_touch` and have other `wl_touch` related code.

So I'm not sure how touch works in X11, but checking if the client binds `wl_touch` shouldn't change anything.

Reply:
Oh interesting! That is new to me, but pretty cool.

Reply:
Presumably it needs `wl_touch` to generate the multi-touch events specified in Xinput 2.2: https://gitlab.freedesktop.org/xorg/proto/xorgproto/-/blob/master/specs/XI2proto.txt.

That protocol also has a section titled "Pointer emulation from multitouch events". Presumably that's handled in XWayland and doesn't require anything from us.